### ISSUE TITLE
[Feature Request] Permissions checker

### ISSUE THREAD

**Issue by keanugrieves** [zenstackhq/zenstack#242]
**Is your feature request related to a problem? Please describe.**
I'd like to be able to check whether user has the permission to perform a create, update or delete operation so I could reflect it on the frontend.

**Describe the solution you'd like**
A "simple" (ha ha) function on each record that optionally takes an input and returns a boolean that indicates whether the user has permission.

**Describe alternatives you've considered**
Rewriting my zmodel permissions in an additional place as a separately callable function.

**Additional context**
Brainstorming the structure, given the model `Foo`:

```
// If user has any create permissions, return true when input not provided or whether they can create with input
foo.canCreate(input?: Prisma.FooUncheckedCreateInput) => boolean;

// If user has update permissions on current record, return true when input not provided or return whether they can update with current record and input
foo.canUpdate(input?: Prisma.FooUncheckedUpdateInput) => boolean;

// If user has delete permissions on current record, return true
foo.canDelete() => boolean;
```

**Comment by gczh** [comment]
@ymc9 adding some questions here:

1. Would the approach to implement this require more database calls or can these all be done at runtime? Ideally, it shouldn't require more database calls for performance reasons.

**Comment by ymc9** [comment]
> @ymc9 adding some questions here:
> 
> 
> 
> 1. Would the approach to implement this require more database calls or can these all be done at runtime? Ideally, it shouldn't require more database calls for performance reasons.

Thanks @gczh . I believe it doesn't involve database calls, however it's possible it'll be served as a backend api so it'll involve an api from the frontend.

We haven't settled the design yet, and I'll share more information when the work starts.

**Comment by Azzerty23** [comment]
Hi @ymc9, I'm interested in working on it. How do you envision the design? Initially, I was wondering if it might be better to create a single generic hook to fetch multiple permissions in a single API call.

To ensure that I could achieve this, I started a draft that I will share here. Hopefully, this will help in the process.

**Comment by ymc9** [comment]
> Hi @ymc9, I'm interested in working on it. How do you envision the design? Initially, I was wondering if it might be better to create a single generic hook to fetch multiple permissions in a single API call.
> 
> To ensure that I could achieve this, I started a draft that I will share here. Hopefully, this will help in the process.

Thank you for starting this work @Azzerty23 ! Another dream feature and I really appreciate it!

Let me go through the draft PR and share my thoughts there.

**Comment by nahtnam** [comment]
Love this idea, it would also be useful for writing tests to ensure that user A can't read user B's records and other similar permission tests. One of the many reasons I prefer zenstack over supabase since RLS is much harder to test

**Comment by ymc9** [comment]
Implemented in 2.1.0
